rfcs/text/0000-application-monitoring-for-ga.md

```diff
@@ -0,0 +1,126 @@
+
+### White-box monitoring
+
+In order to diagnose problems, critical application logic and infrastructure that each system depends on must be monitored. To illustrate, consider an API that depends on AWS S3 and Elasticsearch. If the API goes down and its logs are not informative, there is no way of telling what infrastructure component failed unless there is separate monitoring of each component.
+
+ Use [CloudWatch Metrics](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_concepts.html#Metric), [Stackdriver Metrics](https://cloud.google.com/monitoring/api/metrics), or log-based metrics ([AWS](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CloudWatchLogsConcepts.html), [GCP](https://cloud.google.com/logging/docs/logs-based-metrics/), [Elasticsearch](http://docs.grafana.org/features/datasources/elasticsearch/)) to white-box monitor your critical application logic and infrastructure. 
+ 
+ If you do not know what to monitor in your application, review the [four golden signals](https://landing.google.com/sre/sre-book/chapters/monitoring-distributed-systems/#xref_monitoring_golden-signals). Many of these metrics are provided out-of-the-box by AWS and GCP managed services.
+ 
+ Your monitoring coverage should be as "[simple as possible, no simpler](https://landing.google.com/sre/sre-book/chapters/monitoring-distributed-systems/#as-simple-as-possible-no-simpler-lqskHx)" to diagnose common failure modes of your system. Some common examples are included below.
+ 
+ * Common failure modes of Elasticsearch clusters are related to JVM memory pressure and minimum shard storage space. These are good things to have on a dashboard.
+ * HTTP APIs should have a dashboard summarizing rates of requests by HTTP status class (`2xx`, `3xx`, `4xx`, `5xx`) and request latencies by percentile (commonly 50th, 75th, 95th, 99th). Adding an optional route and method dimensions to each metric will help you pinpoint what is failing when errors occur.
+ * Queues should be monitored for length, size (bytes), consumption rates by status, and age of oldest message.
+ 
+### Alerts
+
+We need alerts (or alarms) in order to be notified of problems in the system.
+
+Configure alerts in the [dcp-monitoring](https://github.com/HumanCellAtlas/dcp-monitoring) repo with a [CloudWatch Alarm](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_concepts.html#CloudWatchAlarms). All black-box Route 53 health checks should be configured with an alert. Optionally, you may choose to set thresholds on some of your white-box monitoring metrics if they are highly-correlated with system failure (e.g. an EC2 node running out of disk space).
+
+Alerts are currently propagated to the Slack channel specified in the CloudWatch Alarm configuration.
+
+### Dashboards
+
```

Comment:
It would be great if, in each of these sections, you discuss an example how one component implemented this feature, and/or provide code examples or pointers to files on GitHub.

Reply:
This was a good suggestion. It helped tie everything together.